File: TS/系统/03．技能系统/00．技能模板+函数/04．机制组件/08．机制触发/11．条件伤害修正.ts
/** @noSelfInFile */

import type { DamageModifierContext } from "../../../../04．伤害系统/00．伤害计算/06．伤害修正回调";
import type { 机制清理篮子 } from "../06．机制清理/01．机制清理篮子";

const { registerDamageModifier, unregisterDamageModifier } = require("系统.04．伤害系统.00．伤害计算.06．伤害修正回调") as {
  registerDamageModifier: (this: void, callback: (this: void, context: DamageModifierContext) => number, priority?: number) => number;
  unregisterDamageModifier: (this: void, id: number) => boolean;
};

export interface 条件伤害修正参数 {
  名称?: string;
  优先级?: number;
  初始启用?: boolean;
  条件: (this: void, context: DamageModifierContext) => boolean;
  修正: (this: void, context: DamageModifierContext) => number;
  清理?: 机制清理篮子;
}

export interface 条件伤害修正控制器 {
  readonly 名称: string;
  是否启用(): boolean;
  设置启用(启用: boolean): void;
  停止(): void;
}

class 条件伤害修正实现 implements 条件伤害修正控制器 {
  readonly 名称: string;
  private readonly 参数: 条件伤害修正参数;
  private 修正ID = 0;
  private 已停止 = false;
  private 已启用: boolean;

  constructor(参数: 条件伤害修正参数) {
    this.参数 = 参数;
    this.名称 = 参数.名称 ?? "条件伤害修正";
    this.已启用 = 参数.初始启用 !== false;
    const self = this;
    this.修正ID = registerDamageModifier(function 条件伤害修正回调(this: void, context: DamageModifierContext): number {
      return self.处理伤害(context);
    }, 参数.优先级 ?? 0);
  }

  是否启用(): boolean {
    return !this.已停止 && this.已启用;
  }

  设置启用(启用: boolean): void {
    if (this.已停止) return;
    this.已启用 = 启用;
  }

  停止(): void {
    if (this.已停止) return;
    this.已停止 = true;
    this.已启用 = false;
    if (this.修正ID !== 0) {
      unregisterDamageModifier(this.修正ID);
      this.修正ID = 0;
    }
  }

  private 处理伤害(context: DamageModifierContext): number {
    const currentDamage = context.currentDamage;
    if (!this.是否启用() || !(currentDamage > 0)) return currentDamage;
    if (!this.参数.条件(context)) return currentDamage;
    const modifiedDamage = this.参数.修正(context);
    return typeof modifiedDamage === "number" && modifiedDamage === modifiedDamage ? modifiedDamage : currentDamage;
  }
}

export function 创建条件伤害修正(this: void, 参数: 条件伤害修正参数): 条件伤害修正控制器 {
  const 控制器 = new 条件伤害修正实现(参数);
  if (参数.清理 != null) {
    参数.清理.登记清理(`${控制器.名称}-清理`, function 条件伤害修正清理(this: void): void {
      控制器.停止();
    });
  }
  return 控制器;
}

File: src/系统/03．技能系统/00．技能模板+函数/04．机制组件/08．机制触发/11．条件伤害修正.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local ____exports = {}
local ____require_result_0 = require("系统.04．伤害系统.00．伤害计算.06．伤害修正回调")
local registerDamageModifier = ____require_result_0.registerDamageModifier
local unregisterDamageModifier = ____require_result_0.unregisterDamageModifier
local _____6761_4EF6_4F24_5BB3_4FEE_6B63_5B9E_73B0 = __TS__Class()
_____6761_4EF6_4F24_5BB3_4FEE_6B63_5B9E_73B0.name = "条件伤害修正实现"
function _____6761_4EF6_4F24_5BB3_4FEE_6B63_5B9E_73B0.prototype.____constructor(self, _____53C2_6570)
    self["修正ID"] = 0
    self["已停止"] = false
    self["参数"] = _____53C2_6570
    self["名称"] = _____53C2_6570["名称"] or "条件伤害修正"
    self["已启用"] = _____53C2_6570["初始启用"] ~= false
    local ____self = self
    self["修正ID"] = registerDamageModifier(
        function(context)
            return ____self["处理伤害"](____self, context)
        end,
        _____53C2_6570["优先级"] or 0
    )
end
_____6761_4EF6_4F24_5BB3_4FEE_6B63_5B9E_73B0.prototype["是否启用"] = function(self)
    return not self["已停止"] and self["已启用"]
end
_____6761_4EF6_4F24_5BB3_4FEE_6B63_5B9E_73B0.prototype["设置启用"] = function(self, _____542F_7528)
    if self["已停止"] then
        return
    end
    self["已启用"] = _____542F_7528
end
_____6761_4EF6_4F24_5BB3_4FEE_6B63_5B9E_73B0.prototype["停止"] = function(self)
    if self["已停止"] then
        return
    end
    self["已停止"] = true
    self["已启用"] = false
    if self["修正ID"] ~= 0 then
        unregisterDamageModifier(self["修正ID"])
        self["修正ID"] = 0
    end
end
_____6761_4EF6_4F24_5BB3_4FEE_6B63_5B9E_73B0.prototype["处理伤害"] = function(self, context)
    local currentDamage = context.currentDamage
    if not self["是否启用"](self) or not (currentDamage > 0) then
        return currentDamage
    end
    if not self["参数"]["条件"](context) then
        return currentDamage
    end
    local modifiedDamage = self["参数"]["修正"](context)
    return type(modifiedDamage) == "number" and modifiedDamage == modifiedDamage and modifiedDamage or currentDamage
end
____exports["创建条件伤害修正"] = function(_____53C2_6570)
    local _____63A7_5236_5668 = __TS__New(_____6761_4EF6_4F24_5BB3_4FEE_6B63_5B9E_73B0, _____53C2_6570)
    if _____53C2_6570["清理"] ~= nil then
        local ____self_1 = _____53C2_6570["清理"]
        ____self_1["登记清理"](
            ____self_1,
            _____63A7_5236_5668["名称"] .. "-清理",
            function()
                _____63A7_5236_5668["停止"](_____63A7_5236_5668)
            end
        )
    end
    return _____63A7_5236_5668
end
return ____exports
